### PR TITLE
Refactor auth tests

### LINKED ISSUE #1: Modularize the auth extension tests
As part of a larger move the refactor the auth extension code, I'd like to start by breaking up the ~6500 LOC test suite for the auth extension into a number of smaller suites with more focus setup and tear down. I hope to add _many more_ tests, but the only way to reasonably do that is to break it up a bit, and probably prune the tests down.

### PR DESCRIPTION
**Pull request by scotttrinh**
The auth tests were getting very hard to add to, with the original test being ~6500 LOC with a ton of duplication. Split this into several test suites and fixed some of the test running infrastructure needed to run tests by path with relative imports.

I hope to add substantially more tests to the auth extension ahead of a larger refactor, so wanted to get this into a good state to do so.

Closes #1

### CODE REVIEW

File: tests/auth/base.py [hunk 1]
```diff
@@ -0,0 +1,590 @@
+#
+# This source file is part of the EdgeDB open source project.
```
Comment (clarkg) [inline tests/auth/base.py]: lol strictly speaking "Gel"?

**Review by clarkg — APPROVED** [review]
@scotttrinh to be transparent I obviously don't know what I'm talking about, but we really appreciate you doing this and I want to unblock you. I'm also generally of the mindset that all test improvements are good improvements

personally I've tried most of the code review bots (e.g. Graphite, Cursorbot, Greptile) and maybe we could consider doing this in this repo too. but I've actually found them all less effective and lower signal than asking GPT 5.2 (extra high) to critique code, especially when I'm looking at something I'm not familiar with

in fact what I do (I dub this "quad slop" or "council review") is run four models via Cursor's parallel model execution (prompt: "critique @/tests/auth")  and then do a second round to double check and synthesize the reviews from the first round. Obviously token expensive, but I think it's generally high coverage and high signal. the "downside" is AI will probably always find problems with any non-trivial piece of code no matter how many rounds of this you do

anyways here's the output of some "second rounds" (labelled A-D). i thought it was interesting even these second rounds don't agree on priority with each other. i'd be curious what you think! are they coherent or salient, more signal than noise?:

*********
**** A ****
## One synthesized, prioritized critique (what matters most)

### P0 — Test *correctness/meaning* risks (fix first because they can invalidate what you think you’re testing)
- **PKCE param-name mismatch (`code_verifier` vs `verifier`)**: As written, the “wrong verifier” branch can accidentally become a “missing verifier” test if the endpoint only honors one key. Either standardize on the real API key, or add an explicit alias test and keep the negative/positive paths consistent.
- **Email helper returns `(None, None)` silently**: This makes failures look like downstream parsing/`None` errors, not “email wasn’t produced”. Even as a pure helper, it’s better for correctness to fail loudly (with clear diagnostics) once the test expects an email.
- **Overlapping provider secrets**: Not a smoking gun, but it’s a straightforward way to strengthen “correct provider selected” assertions. (Low cost, small but real correctness hardening.)

### P1 — Flakiness / “spooky action at a distance” (highest operational risk)
- **Non-transactional suite requires strict cleanup discipline**: With `TRANSACTION_ISOLATION = False`, any missed reset can bleed into later tests. This is the single biggest structural flake risk.
- **Config propagation timing is handled inconsistently**:
  - Sometimes you wait after a config mutation (webhooks, email provider name).
  - Often you *don’t* wait after restoring defaults in `finally`, and many provider-config flips don’t wait at all.
  - Given the suite shares DB state, the safest pattern is: *every* `CONFIGURE … SET/RESET/INSERT` that affects HTTP behavior should be followed by a deterministic “server has observed it” wait—both in setup and in teardown.
- **Email capture via `/tmp` pickle is inherently cross-test sensitive**: it’s external state, not namespaced to the test process, and not cleaned up. Even if emails are “usually synchronous”, this is a classic sporadic-flake vector.
- **Broad reset of all webhooks in `test_auto_signup`**: It’s correct cleanup but more destructive than necessary; it increases coupling with any future tests that start relying on webhook config remaining intact during the method’s later scenarios.

### P2 — Maintainability / failure diagnosis (high developer-time cost)
- **Scenario bundles (“mega-tests”)**: They amortize setup cost, but they reduce the “one failing behavior → one failing test” property. The best compromise here is often:
  - keep the single method (to avoid setup cost), but
  - wrap scenarios/variants in `self.subTest(...)` (especially matrix loops like magic-link missing-key cases).
- **Helper surface area**:
  - `_test_oauth_callback()` is approaching “mini framework” territory; it’s easy to pass the wrong combination of knobs. A small “flow config” struct/object per provider/flow would make call sites harder to misuse.
  - PKCE generation should come from the canonical helper (`generate_pkce_pair`) to reduce drift and make intent obvious.
- **Readability papercuts**: variable reuse (e.g. reusing `webhook_request` for a different type), and scattered plain `assert` usage reduce debug clarity.

### P3 — Coverage & performance opportunism (nice-to-have, after stability)
- **Long waits on webhook checks** (`timeout=120`) are great for reducing flakes but expensive on failure; the suite could fail faster *with better debug output* (dump observed requests, routes registered, etc.).
- **A few “claimed invariants” aren’t asserted**:
  - e.g. `test_magic_code_flow` comments that non-existent users shouldn’t trigger email, but doesn’t assert “no email written”.
  - `test_identity_cascade_delete` deletes but doesn’t query to confirm children are gone (it may still catch constraint errors, but it’s weaker than it could be).
- **Cookie/security attributes**: tests validate cookie *values*, not flags (Secure/HttpOnly/SameSite). Whether that matters depends on what regressions you care about.

If you want, I can turn this into a short “action plan” ordered by ROI (what to change first without re-architecting the suite), but the core bottom line is: **tighten PKCE naming + make config/email handling deterministic** first; everything else is secondary refactor/coverage work.

**** B ****
## Synthesized, prioritized critique of `tests/auth/`

### Highest priority (reliability / flakiness risk)
- **Shared DB/config state + propagation timing**
  - **Why it matters**: with `TRANSACTION_ISOLATION = False`, any missed reset or “reset done but not settled” can create order dependence.
  - **What to do**: standardize a pattern like “mutate config → wait until visible → run assertions → restore config → wait until visible”, including in `finally` blocks.

- **Email capture helper is brittle**
  - **Current shape**: reads a pickle in `/tmp`, returns `(None, None)` when missing, and parses HTML with regex:

```489:518:/Users/clarkg/.cursor/worktrees/gel/eyv/tests/auth/base.py
def _verify_email_file(self, email):
    file_name_hash = hashlib.sha256(f"{SENDER}{email}".encode()).hexdigest()
    test_file = os.environ.get(
        "EDGEDB_TEST_EMAIL_FILE",
        f"/tmp/edb-test-email-{file_name_hash}.pickle",
    )
    if not os.path.exists(test_file):
        return None, None
    ...
    match_link = re.search(
        r'<p style="word-break: break-all">([^<]+)', html_email
    )
    ...
    match_code = re.search(r'(?:^|\s)(\d{6})(?:\s|$)', html_email)
    ...
    return link, code
```

  - **Why it matters**: easy to become racy if email write becomes async-ish; regex couples tests to presentation; returning `None` can lead to confusing downstream errors.
  - **What to do**: add polling with timeout + better failure messages; extract URLs/codes in a way less coupled to exact HTML; optionally delete the file after reading to prevent stale artifacts.

### Medium priority (maintainability / debuggability)
- **Multi-scenario bundle tests reduce failure isolation**
  - **Examples**: `test_magic_link_flow` (3 flows), `test_auto_signup` (3 modes + broad resets), `test_register_form_flow` (form+json+validation), `test_otc_mechanics` (multiple “otc_*” cases).
  - **What to do**: split into separate tests *or* keep one method but use `self.subTest("scenario name")` so you get “one failing behavior == one failing subcase”.

- **PKCE naming drift / ambiguity**
  - **Why it matters**: the `code_verifier` vs `verifier` mix in `test_pkce_token_exchange` makes intent unclear and can silently miss the intended “wrong verifier” coverage.
  - **What to do**: pick one canonical parameter name in tests; if the API supports aliases, add an explicit alias-compat test instead of mixing names inside a single flow.

- **OAuth helper API is “too wide”**
  - `_test_oauth_callback` has many optional knobs; it works, but increases cognitive load and risk of misuse.
  - **What to do**: group parameters into a small config object per flow/provider (even just a dict/dataclass) so call sites become declarative.

### Lower priority (test strength / clarity / minor brittleness)
- **Provider secrets should be distinct** (to better detect “wrong provider selected” bugs).
- **Replace bare `assert` with `self.assert*`** for clearer failures and to avoid `-O` hazards.
- **Cookie parsing helper is simplistic** (single `set-cookie`, first cookie only). If you start asserting flags, you’ll likely need a more robust parse than:

```475:484:/Users/clarkg/.cursor/worktrees/gel/eyv/tests/auth/base.py
set_cookie = headers.get("set-cookie")
if set_cookie is not None:
    (k, v) = set_cookie.split(";", 1)[0].split("=", 1)
```

- **Fill a few “comment-only” assertions** (e.g. “no email sent for nonexistent user”) and strengthen cascade-delete checks by querying post-delete state.
- **Tight presentation assertions** (e.g. exact `rp.name == f"{APP_NAME[:100]}..."`) are brittle unless you truly want to lock formatting; consider asserting the semantic property instead.

**** C ****
## One synthesized, prioritized critique (thorough but ordered by ROI/risk)

### 1) Flakiness / order-dependence risks (highest priority)
- **Email capture is a future flake magnet**: `_verify_email_file()` does a single `exists()` check (no waiting), depends on `/tmp` state, and parses HTML via a regex that encodes a specific template fragment. This is exactly the kind of helper that breaks when delivery becomes more async or templates change.
- **Config propagation is only half-disciplined**: many tests correctly `_wait_for_db_config(...)` after mutating config that affects HTTP behavior, but often do **not** wait after restoring defaults in `finally` blocks (example above). Given `TRANSACTION_ISOLATION=False`, this is the most plausible source of “spooky action at a distance”.
- **Over-broad cleanup**: `RESET ext::auth::WebhookConfig` (no filter) is a big hammer and increases coupling risk (especially if you ever change suite granularity/parallelism).
- **Mock server lifecycle edge-case**: starting mock servers before `super().setUp()` can leak threads on setup failure.

### 2) Failure isolation & debuggability (high priority)
- **Mega-tests reduce signal**: several tests are “scenario bundles” (e.g. `test_auto_signup`, `test_magic_link_flow`, `test_register_form_flow`, `test_otc_mechanics`). When one sub-invariant fails, you get one failing test with lots of unrelated steps.
  - Best compromise: keep the single test (amortize setup) but wrap each scenario in `self.subTest("...")`.
- **Webhook waits are expensive on failure**: `try_until_succeeds(delay=2, timeout=120, ...)` is pragmatic, but 120s timeouts make failures slow and noisy unless the helper also dumps “what requests did we see?” on timeout.

### 3) Maintainability / helper hygiene (medium priority)
- **PKCE generation is duplicated** even though `generate_pkce_pair()` exists; it’s easy for these to drift or confuse readers.
- **OAuth “helper framework” is hard to evolve**: `_test_oauth_callback()` is very parameter-heavy; it works, but it’s easy to misuse and hard to refactor.
- **Cookie parsing helper is fragile**: `maybe_get_cookie_value()` assumes a single `set-cookie` header and only parses the first cookie in it; it will break if the server starts setting multiple cookies or reorders them.

### 4) Test strength / coverage nits (lower priority but real)
- **Non-unique provider secrets weaken assertions** (Google/Azure/Apple share; Discord/Slack share), reducing the chance of catching “wrong provider config used”.
- **“Claims without assertions” exist**: e.g. `test_magic_code_flow` comments “Should NOT send email for non-existent user” but doesn’t assert that; `test_identity_cascade_delete` doesn’t verify dependent rows are gone after delete.
- **Presentation-coupled assertions** (like exact truncation) are fine if intentional, but otherwise they make tests brittle without improving security/correctness.

### 5) Small hygiene issues (nice-to-have)
- **Unused `argon2` hasher** in `tests/auth/base.py`.
- **Variable re-use harms readability** (e.g., `webhook_request` used as both the request-key tuple and later as a request object in `test_core.py`).
- **Plain `assert` usage** in tests is a minor style smell (less informative failures; disabled under `-O`).

**** D ****
### Synthesized critique (thorough but prioritized)

- **P0 — Correctness & flake risk (highest priority)**
  - **Config propagation isn’t consistently “waited-for” (including after restore).** With `TRANSACTION_ISOLATION = False`, a missing `_wait_for_db_config(...)` after restoring defaults can create order-dependent failures within a class (B’s example is real in `test_register_no_smtp`).
  - **Email capture is brittle and potentially racy.** `_verify_email_file()` does “check once, then parse HTML via regex” and returns `(None, None)` if the file isn’t there yet; many tests read it immediately after the triggering request.
  - **PKCE verifier key mismatch (`code_verifier` vs `verifier`) is a real “might be testing the wrong behavior” issue.** It’s the only `code_verifier` usage in `tests/auth/`; everything else uses `verifier`.
  - **Very long webhook polling timeouts (120s) make failures slow and opaque.** Reliability is good, but when something breaks you can burn minutes per test without great debug output.
  - **`setUp()` mock-server lifetime hazard.** Starting servers before `super().setUp()` can leak on `setUp` failure (rare, but real).

- **P1 — Failure isolation & maintainability**
  - **Multi-scenario “mega-tests” reduce signal.** You have several “flow bundles” (magic-link, email-password register flow, PKCE token exchange, etc.). Using `subTest(...)` per scenario is a good compromise when you want to share setup.
  - **The OAuth helper is good *and* too wide.** `_test_oauth_callback()` is powerful but has many knobs; a small config object/builder would make call sites harder to misuse and easier to read.
  - **Helper drift/duplication.** PKCE generation is duplicated even though `ExtAuthTestCase.generate_pkce_pair()` exists; cookie parsing is ad-hoc; request-shape handling (JSON vs form) is re-solved in places.

- **P2 — Test strength / missing assertions**
  - **Duplicate provider secrets weaken sensitivity** (several providers share the same secret). It won’t hide most “wrong host/path” bugs, but it *can* hide “wrong secret chosen” bugs.
  - **“No email for nonexistent user” is mentioned but not asserted** in `test_magic_code_flow`.
  - **Cascade delete test is “no error on delete” rather than “assert orphans gone”.** It still catches some regressions, but it’s weaker than it could be.
  - **Presentation-coupled assertions** (e.g. exact truncation `f"{APP_NAME[:100]}..."` in UI/WebAuthn) are brittle unless you explicitly want to lock that behavior.

- **P3 — Hygiene**
  - **Unused `argon2.PasswordHasher()` in `tests/auth/base.py`.**
  - **A few bare `assert` statements** (fine under normal test runs, but less informative than `self.assert*` and technically elided under `python -O`).

If you want, I can also rank *which single refactor* would pay off most: it’s **a tiny “configure + wait + restore + wait” context helper** plus **a “wait_for_email(timeout)” helper**—those two address most of the real flake vectors without changing test intent.

File: tests/auth/base.py (hunk 1 above)
Comment (scotttrinh) [reply tests/auth/base.py]: Yeah, our usage of Gel and EdgeDB in the license headers is pretty inconsistent (as well as the dates). I'd rather do a separate pass where we align those things.

**Comment by scotttrinh** [comment]
> We should look at whether it's worth setting up CI on this repo.

Absolutely. Hopefully even just generic GitHub Actions workflows will be free and useful as a gate even while we're figuring out the asset publishing pipeline.
